src/pipecat/services/tavus.py

```diff
@@ -0,0 +1,135 @@
+        self._replica_id = replica_id
+        self._persona_id = persona_id
+        self._session = session
+
+        self._conversation_id: str
+        # self._room_url: str
+
+    # def get_room_url(self) -> str:
+    #     return self._room_url
+
+    async def initialize(self) -> str:
+        url = "https://tavusapi.com/v2/conversations"
+        headers = {"Content-Type": "application/json", "x-api-key": self._api_key}
+        payload = {
+            "replica_id": self._replica_id,
+            "persona_id": self._persona_id,
+        }
+        async with self._session.post(url, headers=headers, json=payload) as r:
+            r.raise_for_status()
+            response_json = await r.json()
+
+        logger.debug(f"TavusVideoService joined {response_json['conversation_url']}")
+        self._conversation_id = response_json["conversation_id"]
+        return response_json["conversation_url"]
+
+    def can_generate_metrics(self) -> bool:
+        return True
+
+    async def get_persona_name(self) -> str:
+        url = f"https://tavusapi.com/v2/personas/{self._persona_id}"
+        headers = {"Content-Type": "application/json", "x-api-key": self._api_key}
+        async with self._session.get(url, headers=headers) as r:
+            r.raise_for_status()
+            response_json = await r.json()
+
+        logger.debug(f"TavusVideoService persona grabbed {response_json}")
+        return response_json["persona_name"]
+
+    async def end_conversation(self) -> None:
+        url = f"https://tavusapi.com/v2/conversations/{self._conversation_id}/end"
+        headers = {"Content-Type": "application/json", "x-api-key": self._api_key}
+        async with self._session.post(url, headers=headers) as r:
+            r.raise_for_status()
+
+    async def _encode_audio_and_send(self, audio: bytes, done: bool) -> None:
+        """Encodes audio to base64 and sends it to Tavus"""
+        audio_base64 = base64.b64encode(audio).decode("utf-8")
+        logger.trace(f"TavusVideoService sending {len(audio)} bytes")
+        await self._send_audio_message(audio_base64, done=done)
+
+    async def process_frame(self, frame: Frame, direction: FrameDirection):
+        await super().process_frame(frame, direction)
+        if isinstance(frame, TTSStartedFrame):
+            await self.start_processing_metrics()
+            await self.start_ttfb_metrics()
+            self._current_idx_str = str(frame.id)
+        elif isinstance(frame, TTSAudioRawFrame):
+            await self._encode_audio_and_send(frame.audio, done=False)
+        elif isinstance(frame, TTSStoppedFrame):
+            await self._encode_audio_and_send(b"\x00", done=True)
+            await self.stop_ttfb_metrics()
+            await self.stop_processing_metrics()
+        elif isinstance(frame, StartInterruptionFrame):
+            await self._send_interrupt_message()
+        elif isinstance(frame, EndFrame):
+            await self.end_conversation()
```

Comment:
`cleanup()` might be safer, because you can also get a `CancelFrame` if a SIGTERM or SIGINT signals are sent. Or you can call `_end_conversation()` also when you receive a `CancelFrame`.

Reply:
changed it to `elif isinstance(frame, (EndFrame, CancelFrame)):`